File: APE-05 Creación, Lectura y Escritura de Ficheros/Ejercicio6.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

public class Ejercicio6 {

    private static final String NOMBRE_ARCHIVO = "asistencia.txt";

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int opcion;

        do {
            mostrarMenu();
            opcion = leerEntero(scanner, "Seleccione una opción: ");

            switch (opcion) {
                case 1:
                    registrarAsistencia(scanner);
                    break;
                case 2:
                    mostrarHistorial();
                    break;
                case 3:
                    contarFaltas();
                    break;
                case 4:
                    System.out.println("Saliendo del programa. ¡Hasta luego!");
                    break;
                default:
                    System.out.println("Opción inválida. Intente de nuevo.");
            }

            System.out.println();
        } while (opcion != 4);

        scanner.close();
    }

    private static void mostrarMenu() {
        System.out.println("=== Control de Asistencia ===");
        System.out.println("1. Registrar asistencia");
        System.out.println("2. Mostrar historial");
        System.out.println("3. Contar faltas");
        System.out.println("4. Salir");
    }

    private static int leerEntero(Scanner scanner, String mensaje) {
        int valor;
        while (true) {
            System.out.print(mensaje);
            String linea = scanner.nextLine().trim();
            try {
                valor = Integer.parseInt(linea);
                return valor;
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un número válido.");
            }
        }
    }

    private static void registrarAsistencia(Scanner scanner) {
        String nombre = leerTextoNoVacio(scanner, "Nombre del estudiante: ");
        String fecha = leerTextoNoVacio(scanner, "Fecha (dd/mm/aaaa): ");
        String estado = leerEstado(scanner);

        String registro = nombre + "|" + fecha + "|" + estado;
        guardarRegistro(registro);

        System.out.println("Asistencia registrada correctamente.");
    }

    private static String leerTextoNoVacio(Scanner scanner, String mensaje) {
        String texto;
        do {
            System.out.print(mensaje);
            texto = scanner.nextLine().trim();
            if (texto.isEmpty()) {
                System.out.println("El valor no puede estar vacío. Intente de nuevo.");
            }
        } while (texto.isEmpty());
        return texto;
    }

    private static String leerEstado(Scanner scanner) {
        String estado;
        do {
            System.out.print("Estado (P = presente, F = falta): ");
            estado = scanner.nextLine().trim().toUpperCase();
            if (!estado.equals("P") && !estado.equals("F")) {
                System.out.println("Estado inválido. Ingrese P o F.");
            }
        } while (!estado.equals("P") && !estado.equals("F"));
        return estado;
    }

    private static void guardarRegistro(String registro) {
        File archivo = new File(NOMBRE_ARCHIVO);
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(archivo, true))) {
            writer.write(registro);
            writer.newLine();
        } catch (IOException e) {
            System.out.println("Error al guardar el registro: " + e.getMessage());
        }
    }

    private static void mostrarHistorial() {
        File archivo = new File(NOMBRE_ARCHIVO);
        if (!archivo.exists()) {
            System.out.println("No hay historial de asistencia registrado.");
            return;
        }

        System.out.println("--- Historial de asistencia ---");
        try (BufferedReader reader = new BufferedReader(new FileReader(archivo))) {
            String linea;
            int contador = 0;
            while ((linea = reader.readLine()) != null) {
                contador++;
                String[] partes = linea.split("\\|");
                if (partes.length == 3) {
                    String nombre = partes[0];
                    String fecha = partes[1];
                    String estado = partes[2].equals("P") ? "Presente" : "Falta";
                    System.out.printf("%d. %s - %s - %s%n", contador, nombre, fecha, estado);
                } else {
                    System.out.printf("%d. Registro inválido: %s%n", contador, linea);
                }
            }
            if (contador == 0) {
                System.out.println("No hay registros en el archivo.");
            }
        } catch (IOException e) {
            System.out.println("Error al leer el historial: " + e.getMessage());
        }
    }

    private static void contarFaltas() {
        File archivo = new File(NOMBRE_ARCHIVO);
        if (!archivo.exists()) {
            System.out.println("No hay registros disponibles para contar faltas.");
            return;
        }

        int faltas = 0;
        try (BufferedReader reader = new BufferedReader(new FileReader(archivo))) {
            String linea;
            while ((linea = reader.readLine()) != null) {
                String[] partes = linea.split("\\|");
                if (partes.length >= 3 && partes[2].equalsIgnoreCase("F")) {
                    faltas++;
                }
            }
            System.out.println("Total de faltas registradas: " + faltas);
        } catch (IOException e) {
            System.out.println("Error al contar faltas: " + e.getMessage());
        }
    }
}

File: APE-05 Creación, Lectura y Escritura de Ficheros/Ejercicio 6.cpp
#include <fstream>
#include <iostream>
#include <limits>
#include <sstream>
#include <string>

const std::string NOMBRE_ARCHIVO = "asistencia.txt";

int leerEntero(const std::string &mensaje) {
    int valor;
    while (true) {
        std::cout << mensaje;
        if (std::cin >> valor) {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            return valor;
        }
        std::cout << "Debe ingresar un numero valido." << std::endl;
        std::cin.clear();
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    }
}

std::string leerTextoNoVacio(const std::string &mensaje) {
    std::string texto;
    do {
        std::cout << mensaje;
        std::getline(std::cin, texto);
        if (texto.empty()) {
            std::cout << "El valor no puede estar vacio. Intente de nuevo." << std::endl;
        }
    } while (texto.empty());
    return texto;
}

std::string leerEstado() {
    std::string estado;
    do {
        std::cout << "Estado (P = presente, F = falta): ";
        std::getline(std::cin, estado);
        for (auto &c : estado) c = toupper(static_cast<unsigned char>(c));
        if (estado != "P" && estado != "F") {
            std::cout << "Estado invalido. Ingrese P o F." << std::endl;
        }
    } while (estado != "P" && estado != "F");
    return estado;
}

void guardarRegistro(const std::string &registro) {
    std::ofstream archivo(NOMBRE_ARCHIVO, std::ios::app);
    if (!archivo) {
        std::cout << "Error al abrir el archivo para guardar el registro." << std::endl;
        return;
    }
    archivo << registro << std::endl;
}

void registrarAsistencia() {
    std::string nombre = leerTextoNoVacio("Nombre del estudiante: ");
    std::string fecha = leerTextoNoVacio("Fecha (dd/mm/aaaa): ");
    std::string estado = leerEstado();
    std::string registro = nombre + "|" + fecha + "|" + estado;
    guardarRegistro(registro);
    std::cout << "Asistencia registrada correctamente." << std::endl;
}

void mostrarHistorial() {
    std::ifstream archivo(NOMBRE_ARCHIVO);
    if (!archivo) {
        std::cout << "No hay historial de asistencia registrado." << std::endl;
        return;
    }

    std::cout << "--- Historial de asistencia ---" << std::endl;
    std::string linea;
    int contador = 0;
    while (std::getline(archivo, linea)) {
        contador++;
        std::stringstream ss(linea);
        std::string nombre, fecha, estado;
        if (std::getline(ss, nombre, '|') && std::getline(ss, fecha, '|') && std::getline(ss, estado)) {
            std::string textoEstado = (estado == "P") ? "Presente" : "Falta";
            std::cout << contador << ". " << nombre << " - " << fecha << " - " << textoEstado << std::endl;
        } else {
            std::cout << contador << ". Registro invalido: " << linea << std::endl;
        }
    }
    if (contador == 0) {
        std::cout << "No hay registros en el archivo." << std::endl;
    }
}

void contarFaltas() {
    std::ifstream archivo(NOMBRE_ARCHIVO);
    if (!archivo) {
        std::cout << "No hay registros disponibles para contar faltas." << std::endl;
        return;
    }

    int faltas = 0;
    std::string linea;
    while (std::getline(archivo, linea)) {
        std::stringstream ss(linea);
        std::string nombre, fecha, estado;
        if (std::getline(ss, nombre, '|') && std::getline(ss, fecha, '|') && std::getline(ss, estado)) {
            if (estado == "F") {
                faltas++;
            }
        }
    }
    std::cout << "Total de faltas registradas: " << faltas << std::endl;
}

void mostrarMenu() {
    std::cout << "=== Control de Asistencia ===" << std::endl;
    std::cout << "1. Registrar asistencia" << std::endl;
    std::cout << "2. Mostrar historial" << std::endl;
    std::cout << "3. Contar faltas" << std::endl;
    std::cout << "4. Salir" << std::endl;
}

int main() {
    int opcion;
    do {
        mostrarMenu();
        opcion = leerEntero("Seleccione una opcion: ");
        switch (opcion) {
            case 1:
                registrarAsistencia();
                break;
            case 2:
                mostrarHistorial();
                break;
            case 3:
                contarFaltas();
                break;
            case 4:
                std::cout << "Saliendo del programa. !Hasta luego!" << std::endl;
                break;
            default:
                std::cout << "Opcion invalida. Intente de nuevo." << std::endl;
        }
        std::cout << std::endl;
    } while (opcion != 4);
    return 0;
}
